File: python/protocol.py
# -*- coding: utf-8 -*-
"""
auto-shutdown —— 通信协议实现（Python 测试服务端用，可移植到 ESP32）

本文件实现与 PC 端（Rust, src/crypto.rs + src/protocol.rs）完全一致的加密帧：

    帧 = Base64( HMAC-SHA256(k_mac, IV || CT)[32B] || IV[16B] || CT )

- IV   : 每条报文随机生成的 16 字节初始向量；
- CT   : AES-256-CTR(k_enc, IV, 明文 JSON 字节)；
- k_enc: SHA-256(PASSPHRASE + ":enc")   —— 加密密钥（32 字节）
- k_mac: SHA-256(PASSPHRASE + ":mac")   —— 校验密钥（32 字节）

注意两端必须使用相同的 PASSPHRASE（默认值见下），否则 HMAC 校验会失败。

## ESP32 移植提示（mbedtls / Arduino）

- AES-256-CTR: mbedtls_aes_context + MBEDTLS_MODE_CTR，
  CTR 计数器为整个 16 字节 IV 按大端整体递增（与 mbedtls、本实现一致）；
- HMAC-SHA256: mbedtls_md_hmac(MBEDTLS_MD_SHA256, k_mac, 32, ...)；
- 密钥派生: mbedtls_sha256 对字符串 "PASSPHRASE:enc" / "PASSPHRASE:mac" 求摘要；
- 随机 IV: esp_fill_random()。

## 报文类型（明文 JSON）

- {"type": "ping",     "ts": <毫秒时间戳>, "nonce": "<hex>"}  PC -> ESP32（心跳）
- {"type": "pong",     "nonce": "<同 ping>", "uptime_s": <秒>} ESP32 -> PC
- {"type": "discover", "nonce": "<hex>"}                     PC 广播（UDP 发现）
- {"type": "announce", "name": "heartbeat-server", "ws_port": 8123}  服务端单播应答（UDP）
"""

import base64
import hashlib
import hmac as hmac_mod
import json
import os

# 与 PC 端 src/protocol.rs 中的 PASSPHRASE 必须一致！
PASSPHRASE = "auto-shutdown-v1"

# 加密帧各部分长度
IV_LEN = 16   # AES 块大小
MAC_LEN = 32  # HMAC-SHA256 输出长度
KEY_LEN = 32  # AES-256 密钥长度

# 由口令派生密钥（模块加载时计算一次）
_KEY_ENC = hashlib.sha256((PASSPHRASE + ":enc").encode("utf-8")).digest()
_KEY_MAC = hashlib.sha256((PASSPHRASE + ":mac").encode("utf-8")).digest()


def _ctr_xor(key: bytes, iv: bytes, data: bytes) -> bytes:
    """AES-256-CTR 加解密（流模式加解密同函数）。

    使用 `cryptography` 库的 CTR 模式：16 字节 IV 作为计数器初值，
    整块按大端递增 —— 与 mbedtls / Rust 端行为一致。
    """
    from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

    cipher = Cipher(algorithms.AES(key), modes.CTR(iv))
    enc = cipher.encryptor()
    return enc.update(data) + enc.finalize()


def _frame_mac(iv: bytes, ct: bytes) -> bytes:
    """计算帧 HMAC（覆盖 IV 与密文）"""
    return hmac_mod.new(_KEY_MAC, iv + ct, hashlib.sha256).digest()


def seal(obj: dict, iv: bytes | None = None) -> str:
    """加密一条 JSON 报文，返回 Base64 文本帧。

    iv 参数仅供测试（固定 IV 得到确定性输出）；正常运行时使用随机 IV。
    """
    plaintext = json.dumps(obj, separators=(",", ":")).encode("utf-8")
    iv = iv or os.urandom(IV_LEN)
    if len(iv) != IV_LEN:
        raise ValueError("IV 必须为 16 字节")
    ct = _ctr_xor(_KEY_ENC, iv, plaintext)
    mac = _frame_mac(iv, ct)
    return base64.b64encode(mac + iv + ct).decode("ascii")


def open_frame(frame: str) -> dict:
    """解密一条 Base64 文本帧，返回其中的 JSON 对象。

    任何解密/校验失败都会抛出 ValueError —— 对调用方而言这意味
    着报文来源不是持有口令的合法设备，直接丢弃即可。
    """
    try:
        data = base64.b64decode(frame.strip())
    except Exception as e:
        raise ValueError(f"报文不是合法的 Base64: {e}") from e
    if len(data) <= MAC_LEN + IV_LEN:
        raise ValueError(f"报文长度不足（{len(data)} 字节）")
    mac, rest = data[:MAC_LEN], data[MAC_LEN:]
    iv, ct = rest[:IV_LEN], rest[IV_LEN:]

    # 先校验 HMAC，再解密（encrypt-then-MAC）
    if not hmac_mod.compare_digest(mac, _frame_mac(iv, ct)):
        raise ValueError("HMAC 校验失败：报文来源不是合法设备")

    plaintext = _ctr_xor(_KEY_ENC, iv, ct)
    try:
        return json.loads(plaintext.decode("utf-8"))
    except Exception as e:
        raise ValueError(f"解密后的报文不是合法 JSON: {e}") from e


def random_nonce(n: int = 8) -> str:
    """生成 2n 位十六进制随机数（用作报文 nonce）"""
    return os.urandom(n).hex()


if __name__ == "__main__":
    # 简单自测：加解密往返 + 固定 IV 的确定性输出（用于与 Rust 端互验）
    msg = {"type": "ping", "ts": 1700000000000, "nonce": "0011223344556677"}
    sealed = seal(msg, iv=bytes(range(16)))
    print("固定 IV 加密帧（供与 Rust 端互验）:")
    print(" ", sealed)
    assert open_frame(sealed) == msg
    print("自测通过：seal/open 往返一致")

File: src/protocol.rs
//! # 通信协议定义
//!
//! 本模块定义 PC 客户端与心跳服务端之间的所有常量与报文格式。
//! 服务端可以用任何 socket 方案实现（Arduino / MicroPython / 任意后端），
//! 需要实现与本文件完全一致的协议。
//!
//! ## 协议概览
//!
//! - **心跳**: PC 作为 WebSocket 客户端，每隔 [`HEARTBEAT_INTERVAL`] 向服务端的
//!   `ws://<ip>:8123/ws` 发送一条加密的 `ping` 报文；服务端立即回复 `pong`。
//!   如果连续 [`HEARTBEAT_TIMEOUT`]（默认 60s）内没有任何一次成功的 ping/pong，
//!   就认为服务端已经离线（典型部署：服务端插在市电上，离线即市电中断），
//!   触发关机倒计时。
//! - **发现**: PC 向局域网广播 UDP 地址 `255.255.255.255:8124` 发送加密的
//!   `discover` 报文；服务端收到后向来源地址单播回复加密的 `announce` 报文。
//!
//! ## 加密帧格式（WebSocket 文本帧 / UDP 数据报，均为 Base64 文本）
//!
//! ```text
//! Base64( HMAC-SHA256(k_mac, IV || CT)[32B] || IV[16B] || CT )
//! ```
//!
//! - `IV`: 每条报文随机生成的 16 字节初始向量；
//! - `CT`: AES-256-CTR(k_enc, IV, 明文 JSON)；
//! - `k_enc` / `k_mac`: 由预设口令（[`PASSPHRASE`]，两端必须一致）派生的
//!   32 字节密钥（SHA-256），分别用于加密和校验。
//!
//! HMAC 先于解密校验（encrypt-then-MAC），任何第三方在不知道口令的情况下
//! 既无法伪造报文，也无法读取/篡改内容，从而避免局域网内其他设备的干扰。
//!
//! 选择 AES-CTR + HMAC-SHA256 是因为 mbedtls（ESP32 等 MCU）与 Python cryptography 都原生支持，
//! Python 侧的 `cryptography` 库也有现成实现，两端移植都非常简单。

// 报文结构体在此项目中主要作为协议文档存在（实际收发用动态 JSON 处理，
// 以便与服务端实现保持最大兼容性）。
#![allow(dead_code)]

use serde::{Deserialize, Serialize};

/// WebSocket 服务端口（服务端监听）
pub const WS_PORT: u16 = 8123;
/// WebSocket 路径
pub const WS_PATH: &str = "/ws";
/// UDP 发现服务端口（服务端监听）
pub const DISCOVERY_PORT: u16 = 8124;

/// 心跳发送间隔（每次心跳都会重新建立 WebSocket 连接，LAN 内开销可忽略）
pub const HEARTBEAT_INTERVAL: std::time::Duration = std::time::Duration::from_secs(3);
/// 心跳超时时间：连续这么长时间没有任何成功的心跳，就认为服务已失联
pub const HEARTBEAT_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(60);

/// 关机倒计时时长（秒）：倒计时结束仍无人响应则自动关机
pub const COUNTDOWN_SECONDS: u64 = 60;
/// “稍后关机”的重新弹出间隔（秒）
pub const SNOOZE_LATER_SECONDS: u64 = 30;

/// 预共享口令。两端（PC 与服务端）必须配置完全一致的内容。
/// 如需更换，请同步修改 python/protocol.py 与服务端代码。
pub const PASSPHRASE: &str = "auto-shutdown-v1";

// ---------------------------------------------------------------------------
// 报文结构（明文 JSON）
// ---------------------------------------------------------------------------

/// 心跳请求：PC -> 服务端
#[derive(Serialize, Deserialize, Debug)]
pub struct Ping {
    /// 报文类型: "ping"
    #[serde(rename = "type")]
    pub typ: String,
    /// 当前时间戳（毫秒），服务端可忽略
    pub ts: u64,
    /// 随机数，服务端需要在 pong 中原样带回
    pub nonce: String,
}

/// 心跳响应：服务端 -> PC
#[derive(Serialize, Deserialize, Debug)]
pub struct Pong {
    /// 报文类型: "pong"
    #[serde(rename = "type")]
    pub typ: String,
    /// 原样返回 ping 中的 nonce
    pub nonce: String,
    /// 服务端已运行的秒数（可选，用于界面展示）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uptime_s: Option<u64>,
    /// 设备身份（ESP 用芯片 ID 的十六进制）。PC 端用它在一台设备
    /// 换了 IP 之后仍能认出是同一台，避免误把别的设备当成目标
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
}

/// 发现请求：PC 广播
#[derive(Serialize, Deserialize, Debug)]
pub struct Discover {
    /// 报文类型: "discover"
    #[serde(rename = "type")]
    pub typ: String,
    /// 随机数
    pub nonce: String,
}

/// 发现响应：服务端单播回复
#[derive(Serialize, Deserialize, Debug)]
pub struct Announce {
    /// 报文类型: "announce"
    #[serde(rename = "type")]
    pub typ: String,
    /// 设备名称（展示用）
    pub name: String,
    /// 设备身份（与 pong 中的 id 一致）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    /// WebSocket 服务端口
    pub ws_port: u16,
    /// 服务端已运行的秒数（可选）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uptime_s: Option<u64>,
}
